pytensor/tensor/rewriting/elemwise.py

```diff
@@ -1243,6 +1294,72 @@ def local_careduce_fusion(fgraph, node):
     "fusion",
     position=10,
 )
+fuse_seqopt.register(
+    "local_inline_composite_constants",
+    in2out(local_inline_composite_constants),
+    "fast_run",
+    "fusion",
+    position=20,
+)
+
+
+@node_rewriter([Elemwise])
+def local_replace_broadcasted_constants(fgraph, node):
+    """Remove broadcasted constants from Elemwise graphs
+
+    Elemwise(matrix, ones((3, 4))) -> Elemwise(vector, ones((1, 1)))
+
+    In cases where the constant influenced the final shape of the Elemwise operation
+    We broadcast (via alloc) the new Elemwise result:
+
+    Elemwise(row, ones((3, 4))) -> Alloc(Elemwise(row, ones((1, 1))), 3, 4)
```

Comment:
In this case the post-rewrite version might actually be slower than the pre-rewrite version.
After the rewrite we add an extra allocation and copy operation. I guess which one is faster in practice depends on how expensive the scalar_op of the Elemwise is...
Maybe we just shouldn't do that rewrite if we'd have to broadcast the output?

Reply:
We can make Alloc become a view so it doesn't re-allocate (or right now, use a BroadcastTo instead of Alloc)

Reply:
We shouldn't have to worry about that at the level of pre-inplace rewrites

Reply:
My point was more that I think we have a whole bunch of logic that just silently assumes that we don't use any weird arrays like `broadcast(np.ones(1), (3,))` at all. Numpy will even just return a read-only array for that, probably to make sure users don't shoot themselves into the foot:

```python
vals = np.ones(1)
bc = np.broadcast_to(vals, (3,))
assert not bc.flags.writeable
```

A later rewrite that would be totally valid with a normal array could turn into something that errors or even gives silently incorrect results.

Let's say we have a graph like

```python
inc_subtensor(pow(ones(3), 1.)[0, 1], 1)
-> inc_subtensor(Alloc(pow(ones(1), 1)), (3,), inplace=False)[0, 1], 1, inplace=False)
-> inc_subtensor(Alloc(pow(ones(1), 1)), (3,), inplace=True)[0, 1], 1, inplace=False)
-> inc_subtensor(Alloc(pow(ones(1), 1)), (3,), inplace=True)[0, 1], 1, inplace=True)
```
The two inplace rewrites seem perfectly innocent in isolation, but because the first inplace rewrite changed how the array is stored, the second one won't work. There is only one storage location for all three entries of the array, so it will return an array like [3, 3, 3] instead of [2, 2, 1].

I think the most sensible solution to this is to just never ever create arrays that refer to the same memory location multiple times. But that means we can't implement `Alloc(ones(1), (3,), inplace=True)` at all.

---

So maybe that rewrite is still fine, and we think that there are more cases where it helps than cases where it hurts (I think that's probably true). But at least we should make sure we disallow `Alloc(ones(1), 3, 4, inplace=True)`.

Reply:
> So maybe that rewrite is still fine, and we think that there are more cases where it helps than cases where it hurts (I think that's probably true). But at least we should make sure we disallow Alloc(ones(1), 3, 4, inplace=True).

I'll add this to the #367 issue

Reply:
After some discussion: We agree that the rewrite is fine, but we shouldn't rewrite the alloc to a broadcast to avoid creating arrays that alias memory. Doing this *might* be fine (there are some checks in the blas wrappers already), but we thought it is not worth the trouble with possible bugs that this could introduce.